core/src/main/java/org/apache/spark/shuffle/sort/LocalShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.shuffle.sort;
+
+import org.apache.spark.network.util.JavaUtils;
+import org.apache.spark.shuffle.IndexShuffleBlockResolver;
+import org.apache.spark.shuffle.api.ShuffleMapOutputWriter;
+import org.apache.spark.shuffle.api.ShufflePartitionChannelWriter;
+import org.apache.spark.shuffle.api.ShufflePartitionWriter;
+
+import java.io.*;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * Writes shuffle output from one shuffle map task to the local file system, with a separate
+ * index & data file.
+ *
+ * This corresponds to the only shuffle storage in spark <= 2.4.
+ */
+public class LocalShuffleMapOutputWriter implements ShuffleMapOutputWriter {
+
+  final IndexShuffleBlockResolver resolver;
+  final int shuffleId;
+  final int mapId;
+  long partitionLengths[];
+
+  File dataTmpFile;
+
+  FileChannel dataTmpFileOutputChannel = null;
+  FileOutputStream dataTmpFileOutputStream = null;
+
+  public LocalShuffleMapOutputWriter(IndexShuffleBlockResolver resolver, int shuffleId, int mapId) {
+    this.resolver = resolver;
+    this.shuffleId = shuffleId;
+    this.mapId = mapId;
+  }
+
+  @Override
+  public ShufflePartitionWriter newPartitionWriter(int partitionId) {
+    return null;
+  }
+
+  @Override
+  public void commitAllPartitions() {
+    // this will delete the temporary index file itself if there is a failure
+    resolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, dataTmpFile);
```

Comment:
How are the partition lengths here set? Perhaps we can have a wrapper around the FileOutputStream/WritableByteChannels to count the bytes as they're written?

Reply:
yeah sorry this isn't all wired together yet ... there is a CountingOutputStream and CountingWritableByteChannel.  There is already some of these, eg. in UnsafeShuffleWriter, to get partition lengths.  I am a little stuck on figuring out how to get ExternalSorter to work at the moment ...